src/PooledArrays.jl

```diff
@@ -377,6 +415,15 @@ function unsafe_pool_push!(pa::PooledArray{T,R}, val) where {T,R}
            )))
     end
     pool_idx = convert(R, _pool_idx)
+    if pa.cow
+        l = pa.lock
+        Threads.lock(l)
+        pa.invpool = copy(pa.invpool)
+        pa.pool = copy(pa.pool)
+        pa.cow = false
```

Comment:
There's a missing piece here: other arrays sharing this pool will still have `cow=true`, even if there's only one array in that situation. So for example once you've taken a slice of an array, adding a new value will always copy the whole pool.

I think this can be fixed by replacing `cow` with an `Atomic{Int}` counting the number of references (that could probably replace the lock, since we're not thread-safe anyway when adding new values). But we also need a finalized to decrement the counter when an array is GCed.

Reply:
Yes - these comments are right (I had the same thoughts after pushing the code). I do not need a lock but just need the count of references present. I will change it exactly this way.